Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+
+namespace FileSystemVisitorLibrary
+{
+    public class FileSystemVisitor
+    {
+        private readonly Func<FileSystemInfo, bool> _filterFunc;
+
+        private readonly IFileSystem _fileSystem;
+
+        private Actions? _action = Actions.Continue;
+
+        public FileSystemVisitor(IFileSystem fileSystem, Func<FileSystemInfo, bool> filterFunc = null)
+        {
+            _fileSystem = fileSystem;
```

Comment:
File system is a critical dependency for this class. Incorrect behavior should be prevented as soon as possible.
```suggestion
            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
```